apm-sniffer/apm-sdk-plugin/spring-plugins/resttemplate-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/resttemplate/async/RestExecuteInterceptor.java

```diff
@@ -44,30 +50,34 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
         String remotePeer = requestURL.getHost() + ":" + (requestURL.getPort() > 0 ? requestURL.getPort() : "https".equalsIgnoreCase(requestURL
             .getScheme()) ? 443 : 80);
 
-        String formatURIPath = requestURL.getPath();
-        AbstractSpan span = ContextManager.createExitSpan(formatURIPath, contextCarrier, remotePeer);
+        String uri = requestURL.getPath();
+        AbstractSpan span = ContextManager.createExitSpan(uri, contextCarrier, remotePeer);
 
         span.setComponent(ComponentsDefine.SPRING_REST_TEMPLATE);
         Tags.URL.set(span, requestURL.getScheme() + "://" + requestURL.getHost() + ":" + requestURL.getPort() + requestURL
             .getPath());
         Tags.HTTP.METHOD.set(span, httpMethod.toString());
         SpanLayer.asHttp(span);
-        Object[] cacheValues = new Object[3];
-        cacheValues[0] = formatURIPath;
-        cacheValues[1] = contextCarrier;
-        objInst.setSkyWalkingDynamicField(cacheValues);
+        RestTemplateRuntimeContextHelper.addUri(uri);
+        RestTemplateRuntimeContextHelper.addContextCarrier(contextCarrier);
     }
 
     @Override
     public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
         Object ret) throws Throwable {
-        Object[] cacheValues = (Object[]) objInst.getSkyWalkingDynamicField();
-        cacheValues[2] = ContextManager.capture();
-        if (ret != null) {
-            String uri = (String) cacheValues[0];
-            ((EnhancedInstance) ret).setSkyWalkingDynamicField(new EnhanceCacheObjects(uri, ComponentsDefine.SPRING_REST_TEMPLATE, SpanLayer.HTTP, (ContextSnapshot) cacheValues[2]));
+        try {
+            ContextSnapshot contextSnapshot = ContextManager.capture();
+            if (ret != null) {
+                String uri = RestTemplateRuntimeContextHelper.getUri();
+                ((EnhancedInstance) ret).setSkyWalkingDynamicField(new EnhanceCacheObjects(uri, ComponentsDefine.SPRING_REST_TEMPLATE, SpanLayer.HTTP, contextSnapshot));
+            }
+        } catch (Throwable e){
+            LOGGER.warn("resttemplate plugin has error: ", e);
+        } finally {
+            ContextManager.stopSpan();
```

Comment:
Move this into the `try`. Because this could have NPE.

Reply:
done